engine/consensus/matching/engine.go

```diff
@@ -462,169 +458,163 @@ func (e *Engine) checkSealing() {
+				Int("required_chunks", requiredChunks).
+				Msg("removing result with invalid number of chunks")
+			continue
+		}
 
-	// request missing execution results, if sealed height is low enough
-	if uint(final.Height-sealed.Height) >= e.requestReceiptThreshold {
-		requestedCount := 0
-		for _, blockID := range missingBlocksOrderedByHeight {
-			if _, ok := missingByBlockID[blockID]; ok {
-				e.requester.EntityByID(blockID, filter.Any)
-				requestedCount++
+		// check that each chunk collected enough approvals
+		allChunksMatched := true
+
+		// TODO: the randomness of chunk assignment for a result is determined
+		// by the result and a block ID.
+		// As a temporary shortcut, we can just use the ID of the executed
+		// block, i.e. blockID = result.BlockID. However, in the full protocol,
+		// blockID is the first block in its fork, which references an Execution
+		// Receipt with an Execution Result identical to result.
+		// (where blockID != result.BlockID)
+		assignment, err := e.assigner.Assign(result, result.BlockID)
+		if state.IsNoValidChildBlockError(err) {
+			continue
+		}
+		if err != nil {
+			return nil, fmt.Errorf("could not assign verifiers: %w", err)
+		}
+
+		// check that each chunk collects enough approvals
+		for _, chunk := range result.Chunks {
+			matched := e.matchChunk(result.ID(), chunk, assignment)
+			if !matched {
+				allChunksMatched = false
+				break
 			}
 		}
-		e.log.Info().
-			Int("count", requestedCount).
-			Msg("requested missing results")
+
+		// add the result to the results that should be sealed
+		if allChunksMatched {
+			e.log.Info().Msg("adding result with sufficient verification")
+			results = append(results, result)
+		}
 	}
 
 	return results, nil
 }
 
-func (e *Engine) sealResult(result *flow.ExecutionResult, final *flow.Header) error {
+// matchChunk checks that the number of ResultApprovals collected by a chunk
+// exceeds the required threshold.
+func (e *Engine) matchChunk(resultID flow.Identifier, chunk *flow.Chunk, assignment *chunks.Assignment) bool {
 
-	// check if we know the block the result pertains to
-	_, err := e.headersDB.ByBlockID(result.BlockID)
-	if errors.Is(err, storage.ErrNotFound) {
-		return errUnknownBlock
-	}
-	if err != nil {
-		return fmt.Errorf("could not check sealed header: %w", err)
-	}
+	// get all the chunk approvals from mempool
+	approvals := e.approvals.ByChunk(resultID, chunk.Index)
 
-	// get the previous result from our mempool or storage
-	previousID := result.PreviousResultID
-	previous, found := e.results.ByID(previousID)
-	if !found {
-		var err error
-		previous, err = e.resultsDB.ByID(previousID)
-		if errors.Is(err, storage.ErrNotFound) {
-			return errUnknownPrevious
-		}
-		if err != nil {
-			return fmt.Errorf("could not get previous result: %w", err)
+	// only keep approvals from assigned verifiers
+	var validApprovers flow.IdentifierList
+	for approverID := range approvals {
+		ok := chmodule.IsValidVerifer(assignment, chunk, approverID)
```

Comment:
There is no check for cryptographic correctness of the approval. For now, we might put invalid approvals in the mempool. We need to ensure that invalid approvals are not counted here.